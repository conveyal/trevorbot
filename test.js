/* globals describe, it */

var expect = require('chai').expect
var nock = require('nock')
var trevorbot = require('./handler').trevorbot

const TIMBUKTU = JSON.stringify({
  location: {
    now: {
      city: 'Timbuktu',
      country: 'Mali'
    }
  }
})

function makeQueryText (text) {
  return {
    body: {
      text: '@trevorbot ' + text,
      user_name: 'someone'
    }
  }
}

describe('trevorbot', () => {
  describe('problems', function () {  // needs to have function so mocha will provide this.timeout?
    this.timeout(6000)
    this.slow(20000)

    it('should handle server connection timeout', (done) => {
      var scope = nock('https://nomadlist.com')
        .get('/trevorgerhardt.json')
        .delayConnection(60000)
        .reply(200, TIMBUKTU)

      trevorbot(makeQueryText('where is Trevor?'), null, (err, result) => {
        scope.done()
        expect(err).to.not.exist
        expect(result).to.have.property('text', 'I couldn\'t figure that out right now :astonished:')
        done()
      })
    })

    it('should handle server response timeout', (done) => {
      var scope = nock('https://nomadlist.com')
        .get('/trevorgerhardt.json')
        .socketDelay(60000)
        .reply(200, TIMBUKTU)

      trevorbot(makeQueryText('where is Trevor?'), null, (err, result) => {
        scope.done()
        expect(err).to.not.exist
        expect(result).to.have.property('text', 'I couldn\'t figure that out right now :astonished:')
        done()
      })
    })
  })

  describe('successes', () => {
    it('unrecognized command should return something', (done) => {
      trevorbot(makeQueryText('Arblegarbl'), null, (err, result) => {
        expect(err).to.not.exist
        expect(result).to.have.property('text', 'I don\'t understand, I\'m afraid :thinking_face:')
        done()
      })
    })

    it('where query for non-trevor person should not work', (done) => {
      trevorbot(makeQueryText('where in the world is Carmen SanDiego?'), null, (err, result) => {
        expect(err).to.not.exist
        expect(result).to.have.property('text', 'I only know where Trevor is.')
        done()
      })
    })

    it('where query for trevor should work', (done) => {
      var scope = nock('https://nomadlist.com')
        .get('/trevorgerhardt.json')
        .reply(200, TIMBUKTU)

      trevorbot(makeQueryText('where is Trevor?'), null, (err, result) => {
        expect(err).to.not.exist
        expect(result).to.have.property('text', 'Trevor is in Timbuktu, Mali (https://nomadlist.com/trevorgerhardt)')
        scope.done()
        done()
      })
    })

    it('trevor should tell a Chuck Norris joke', (done) => {
      var scope = nock('http://api.icndb.com')
        .get('/jokes/random')
        .reply(200, JSON.stringify({
          type: 'success',
          value: {
            id: 554,
            joke: 'Chuck Norris can install a 64 bit OS on 32 bit machines.',
            categories: [
              'nerdy'
            ]
          }
        }))

      trevorbot(makeQueryText('tell me about Chuck Norris'), null, (err, result) => {
        expect(err).to.not.exist
        expect(result).to.have.property('text', 'Chuck Norris can install a 64 bit OS on 32 bit machines.')
        scope.done()
        done()
      })
    })
  })
})
