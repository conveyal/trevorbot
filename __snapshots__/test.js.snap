// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`trevorbot problems should handle server connection timeout 1`] = `
Object {
  "body": "{\\"text\\":\\"I couldn't figure out where Trevor is! :astonished:\\"}",
  "statusCode": 200,
}
`;

exports[`trevorbot problems should handle server response timeout 1`] = `
Object {
  "body": "{\\"text\\":\\"I couldn't figure out where Trevor is! :astonished:\\"}",
  "statusCode": 200,
}
`;

exports[`trevorbot successes trevor should tell a Chuck Norris joke 1`] = `
Object {
  "body": "{\\"text\\":\\"Chuck Norris can install a 64 bit OS on 32 bit machines.\\"}",
  "statusCode": 200,
}
`;

exports[`trevorbot successes unrecognized command should return something 1`] = `
Object {
  "body": "{\\"text\\":\\"I don't understand, I'm afraid :thinking_face:\\"}",
  "statusCode": 200,
}
`;

exports[`trevorbot successes where query for non-trevor person should not work 1`] = `
Object {
  "body": "{\\"text\\":\\"I only know where Trevor and Bob are.\\"}",
  "statusCode": 200,
}
`;

exports[`trevorbot successes where query for trevor should work 1`] = `"Trevor is in Timbuktu, Mali.  The current time there is **. (https://nomadlist.com//@trevorgerhardt.json)"`;
